src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/ResourceContainerRepository.kt

```diff
@@ -0,0 +1,154 @@
+                    relatedDublinCoreIds.forEach { relatedId ->
+                        val ih = ImportHelper(dublinCoreFk, relatedId, dsl)
+                        ih.importCollection(null, rcTree)
+                    }
+                } else {
+                    val ih = ImportHelper(dublinCoreFk, null, dsl)
+                    ih.importCollection(null, rcTree)
+                }
+            }
+        }.subscribeOn(Schedulers.io())
+    }
+
+    private fun linkRelatedResourceContainers(
+            dublinCoreFk: Int,
+            relations: List<String>,
+            dsl: DSLContext)
+            : List<Int> {
+        val relatedIds = mutableListOf<Int>()
+        relations.forEach { relation ->
+            val parts = relation.split('/')
+            resourceMetadataDao.fetchByLanguageAndIdentifier(parts[0], parts[1], dsl)
+                    ?.let { relatedDublinCore ->
+                        // TODO: Only add link if it doesn't exist already
+                        resourceMetadataDao.addLink(dublinCoreFk, relatedDublinCore.id, dsl)
+                        relatedIds.add(relatedDublinCore.id)
+                    }
+        }
+        return relatedIds
+    }
+
+    inner class ImportHelper(val dublinCoreId: Int, val relatedBundleDublinCoreId: Int?, val dsl: DSLContext, var collectionsImported: Int = 0) {
+
+        fun importCollection(parentId: Int?, node: Tree) {
+            (node.value as? Collection)?.let { collection ->
+                when (relatedBundleDublinCoreId) {
+                    null -> addCollection(collection, parentId)
+                    else -> findCollectionId(collection, relatedBundleDublinCoreId)
+                }
+            }.let { collectionId ->
+                // TODO: If we don't find a corresponding collection, we continue on, passing null to collectionId.
+                // TODO ... Eventually, contents will not be created if there is no parentId. This will happen for front
+                // TODO ... matter until we have another solution.
+                for (childNode in node.children) {
+                    importNode(collectionId, childNode)
+                }
+            }
+            collectionsImported++
+            if (collectionsImported % 25 == 0) {
+                println("Collections imported: $collectionsImported")
+            }
+        }
+
+        private fun findCollectionId(collection: Collection, containerId: Int): Int? =
+                collectionDao.fetchBySlugAndContainerId(collection.slug, containerId)
+                        ?.id
+
+        private fun addCollection(collection: Collection, parentId: Int?): Int {
+            val entity = CollectionMapper().mapToEntity(collection).apply {
+                parentFk = parentId
+                dublinCoreFk = dublinCoreId
+            }
+            return collectionDao.insert(entity, dsl)
+        }
+
+        private fun importNode(parentId: Int?, node: TreeNode) {
+            when (node) {
```

Comment:
I've been _almost_ commenting on this line for days.  I can't rationally justify it, but I can't shake the feeling that this `when` should check the type of `node.value` for Collection vs Content instead.